### DISCUSSION TITLE
How to configure socks proxy on KubernetesClient?

### DISCUSSION THREAD

**Discussion by syr** [fabric8io/kubernetes-client#4938] [Q&A]
Hi, 
using io.fabric8:kubernetes-client:6.3.1 and only have seen http/https proxy configs but not for socks proxy

neither in io.fabric8.kubernetes.client.Config nor in io.fabric8.kubernetes.client.http.HttpClient.Builder#proxyAddress (protocol cannot be specified)

```
        Config config = new Config();
        config.setHttpProxy("");
        config.setHttpsProxy("");

        KubernetesClient client = new KubernetesClientBuilder()
                .withConfig(config)
                .withHttpClientBuilderConsumer(builder -> builder.proxyAddress(new InetSocketAddress("myhost", 9999))
        ).build();
```

Or is there some sample code how to build an okHttpClient object and how to define it to be used by KubernetesClient?

**Comment by shawkins** [comment]
> but not for socks proxy

That is correct, the proxy type is not currently a config value.

> Or is there some sample code how to build an okHttpClient object and how to define it to be used by KubernetesClient?

I think what you are trying to do is achievable with each of the client types, you aren't restricted to okhttp.  

Both okhttp and jdk will rely upon system variables to setup the proxy selector - https://docs.oracle.com/javase/8/docs/api/java/net/doc-files/net-properties.html#Proxies - the only issue being is that both implementations will currently set the proxy on their respective builders to NO_PROXY if a proxy address is not specified.  You would have to inhibit that or set the proxy to null - see the XXXHttpClientFactory.addtionalConfig method for access to the underlying client builder.

```
KubernetesClient client = new KubernetesClientBuilder()
        .withConfig(config)
        .withHttpClientFactory(new OkHttpClientFactory() {
             
          @Override
          protected void additionalConfig(Builder builder) {
            builder.proxy(null); // use null for system properties, or you can construct the appropriate proxy or use the proxySelector method
          }
            
        }).build();
```

**Comment by syr** [comment]
Hi @shawkins, thanks for the hint. I tried as you suggested above, but with 
```
builder.proxy(new Proxy(Proxy.Type.SOCKS, new InetSocketAddress("localhost",8888)))
```

But when using the built `client`, it does not care about it, proxy is not set, e.g. `kubernetesClient.httpClient.httpClient.proxy=DIRECT`


P.S. before DefaultKubernetesClient was deprecated, it was simply working by
```
new DefaultKubernetesClient(		
	new OkHttpClientImpl(
		new OkHttpClient().newBuilder()
		.proxy(new Proxy(Proxy.Type.SOCKS, new InetSocketAddress("localhost",8888)))
		.build()
	), config);
```

**Reply by shawkins** [reply]
It's a problem of when additionalConfig is applied.  That's unfortunately happening before the config logic is setting the proxy.  Only the JDK client is in the proper place to do what is described.

It will take an enhancement to re-arrange how the underlying builder is exposed.

**Comment by syr** [comment]
Just checked with jdk http client...seems also not support socks proxy..there is a open ticket just from 2018 (wow)
https://bugs.openjdk.org/browse/JDK-8214516

So I stick to my deprecated approach which at least works.

**Reply by shawkins** [reply]
Can you open / convert this discussion into an issue?  We should discuss whether forcing "no proxy" makes sense, whether to add proxy type to the configuration, and to move some of the configuration logic for okhttp to make overriding easier.

**Reply by shawkins** [reply]
logged as #5086 and reopened #1335
